cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+                        const MaterialModel::MaterialModelOutputs<dim> &out,
+                        const FEValues<dim> &,
+                        const LinearAlgebra::BlockVector &,
+                        std::vector<double> &output) override
+        {
+          const MaterialModel::UnscaledViscosityAdditionalOutputs<dim> *unscaled_viscosity_outputs
+            = out.template get_additional_output<const MaterialModel::UnscaledViscosityAdditionalOutputs<dim>>();
+
+          Assert(unscaled_viscosity_outputs != nullptr,ExcInternalError());
+
+          for (unsigned int q=0; q<output.size(); ++q)
+            output[q] = unscaled_viscosity_outputs->output_values[0][q];
+        }
+    };
+  }
+
+  namespace MaterialModel
+  {
+    template <int dim>
+    void
+    EquilibriumGrainSize<dim>::initialize()
+    {
+      // Get reference viscosity profile from the ascii data
+      if (use_depth_dependent_viscosity)
+        reference_viscosity_coordinates = reference_viscosity_profile->get_interpolation_point_coordinates();
+
+      // Get column index for density scaling
+      if (use_depth_dependent_rho_vs)
+        {
+          rho_vs_depth_profile.initialize(this->get_mpi_communicator());
+          density_scaling_index = rho_vs_depth_profile.get_column_index_from_name("density_scaling");
+        }
+
+      if (use_depth_dependent_thermal_expansivity)
+        {
+          thermal_expansivity_profile.initialize(this->get_mpi_communicator());
+          thermal_expansivity_column_index = thermal_expansivity_profile.get_column_index_from_name("thermal_expansivity");
+        }
+
+      if (use_depth_dependent_dT_vs)
+        {
+          dT_vs_depth_profile.initialize(this->get_mpi_communicator());
+          temperature_scaling_index = dT_vs_depth_profile.get_column_index_from_name("temperature_scaling");
+        }
+
+      // Get column for crustal depths
+      std::set<types::boundary_id> surface_boundary_set;
+      surface_boundary_set.insert(this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"));
+      crustal_boundary_depth.initialize(surface_boundary_set, 1);
+
+      this->get_signals().post_stokes_solver.connect([&](const SimulatorAccess<dim> &,
+                                                         const unsigned int ,
+                                                         const unsigned int ,
+                                                         const SolverControl &,
+                                                         const SolverControl &)
+      {
+        this->update();
+      });
+
+      this->get_signals().post_advection_solver.connect([&](const SimulatorAccess<dim> &,
+                                                            const unsigned int ,
+                                                            const unsigned int ,
+                                                            const SolverControl &)
+      {
+        this->update();
+      });
```

Comment:
Can you explain in a comment why you want the `update()` function to run at this point (i.e. post-stokes- and post-advection-solve)?

Maybe more generally, we've generally tried to follow a workflow where every time step is computed based on the previous timesteps' results, rather than relying on the order in which the different parts of the current time step are solved. As a consequence, I think it would be nice *not* to update the viscosity between the parts of the current time step, unless there is a good reason for it.

Reply:
Separately, you should probably capture the return object of the `connect()` call and store it in a `boost::signals2::connection` variable so that the connection can be disconnected once the material model object is destroyed.

Reply:
The update() function updates the profile that stores the laterally averaged viscosity. This is needed to compute the viscosity in the material model (since it is rescaled so that the lateral average matches the reference profile). Since the viscosity depends on both temperature and velocity, we want to update the lateral average profile after each temperature and Stokes solve. 

(@alarshi you can add this as a comment)